File: Cpp/Medium/47. 全排列 II.cpp
给定一个可包含重复数字的序列，返回所有不重复的全排列。

示例:

输入: [1,1,2]
输出:
[
  [1,1,2],
  [1,2,1],
  [2,1,1]
]

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        if(nums.size() == 0) {
            return res;
        }
        vector<int> curr;
        vector<bool> isVisited(nums.size(), false);
        sort(nums.begin(), nums.end());
        dfs(res, curr, nums, isVisited);
        return res;
    }
    void dfs(vector<vector<int>>& res, vector<int> curr, vector<int> nums, vector<bool> isVisited) {
        if(curr.size() == nums.size()) {
            res.push_back(curr);
            return;
        }
        for(int i = 0; i < nums.size(); i++) {
            if(isVisited[i] == false) {
                isVisited[i] = true;
                curr.push_back(nums[i]);
                dfs(res, curr, nums, isVisited);
                isVisited[i] = false;
                curr.pop_back();
                while(i < nums.size() - 1 && nums[i] == nums[i + 1]) {
                    ++i;
                }
            }
        }
    }
};

File: Java/medium/47. 全排列 II.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> resList = new ArrayList<>();
        Arrays.sort(nums);
        boolean[] hasVisited = new boolean[nums.length];
        backtrack(resList, res, hasVisited, nums);
        return res;
    }
    private void backtrack(List<Integer> resList, List<List<Integer>> res, boolean[] visited, final int[] nums) {
        if(resList.size() == nums.length) {
            res.add(new ArrayList<>(resList));
            return ;
        }
        for(int i = 0; i < visited.length; i++) {
            if(i != 0 && nums[i] == nums[i - 1] && !visited[i - 1]) {
                continue;
            }
            if(visited[i]) {
                continue;
            }
            visited[i] = true;
            resList.add(nums[i]);
            backtrack(resList, res, visited, nums);
            resList.remove(resList.size() - 1);
            visited[i] = false;
        }
    }
}
